File: skills/agent-patterns/templates/csharp/Parallelization.cs
/**
 * Parallelization Pattern Implementation for C#
 * Concurrent LLM calls for independent subtasks
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Anthropic.SDK;
using Anthropic.SDK.Messaging;

namespace AgentPatterns
{
    /// <summary>
    /// Result of a parallel subtask execution
    /// </summary>
    public class SubtaskResult<T>
    {
        public string SubtaskName { get; set; } = string.Empty;
        public T? Result { get; set; }
        public bool Success { get; set; }
        public string? Error { get; set; }
        public TimeSpan Duration { get; set; }
    }

    /// <summary>
    /// Sectioning Parallelizer - divides task into independent subtasks
    /// 
    /// Example:
    /// <code>
    /// var parallelizer = new SectioningParallelizer(client);
    /// var result = await parallelizer.ProcessCodeReviewAsync(code);
    /// </code>
    /// </summary>
    public class SectioningParallelizer
    {
        private readonly AnthropicClient _client;
        private readonly string _model;

        public SectioningParallelizer(
            AnthropicClient client,
            string model = "claude-sonnet-4-20250514")
        {
            _client = client;
            _model = model;
        }

        /// <summary>
        /// Execute multiple subtasks in parallel
        /// </summary>
        public async Task<List<SubtaskResult<string>>> ExecuteParallelAsync(
            List<(string name, string prompt)> subtasks)
        {
            var tasks = subtasks.Select(async subtask =>
            {
                var startTime = DateTime.UtcNow;
                try
                {
                    var messages = new List<Message>
                    {
                        new Message(RoleType.User, subtask.prompt)
                    };

                    var parameters = new MessageParameters
                    {
                        Messages = messages,
                        Model = _model,
                        MaxTokens = 2048
                    };

                    var response = await _client.Messages.GetClaudeMessageAsync(parameters);
                    var result = response.Content.First().Text;

                    return new SubtaskResult<string>
                    {
                        SubtaskName = subtask.name,
                        Result = result,
                        Success = true,
                        Duration = DateTime.UtcNow - startTime
                    };
                }
                catch (Exception ex)
                {
                    return new SubtaskResult<string>
                    {
                        SubtaskName = subtask.name,
                        Success = false,
                        Error = ex.Message,
                        Duration = DateTime.UtcNow - startTime
                    };
                }
            }).ToList();

            return (await Task.WhenAll(tasks)).ToList();
        }

        /// <summary>
        /// Code review with parallel analysis
        /// </summary>
        public async Task<CodeReviewResult> ProcessCodeReviewAsync(string code)
        {
            var subtasks = new List<(string name, string prompt)>
            {
                ("security", $@"Analyze this code for security vulnerabilities:
```
{code}
```
List any security issues found with severity and recommendations."),

                ("performance", $@"Analyze this code for performance issues:
```
{code}
```
Identify inefficiencies and suggest optimizations."),

                ("maintainability", $@"Analyze this code for maintainability:
```
{code}
```
Check code structure, naming, and suggest improvements."),

                ("bugs", $@"Analyze this code for potential bugs:
```
{code}
```
Identify logic errors, edge cases, and potential runtime issues.")
            };

            var results = await ExecuteParallelAsync(subtasks);

            return new CodeReviewResult
            {
                SecurityAnalysis = results.FirstOrDefault(r => r.SubtaskName == "security")?.Result ?? "",
                PerformanceAnalysis = results.FirstOrDefault(r => r.SubtaskName == "performance")?.Result ?? "",
                MaintainabilityAnalysis = results.FirstOrDefault(r => r.SubtaskName == "maintainability")?.Result ?? "",
                BugAnalysis = results.FirstOrDefault(r => r.SubtaskName == "bugs")?.Result ?? "",
                TotalDuration = results.Max(r => r.Duration)
            };
        }
    }

    public class CodeReviewResult
    {
        public string SecurityAnalysis { get; set; } = string.Empty;
        public string PerformanceAnalysis { get; set; } = string.Empty;
        public string MaintainabilityAnalysis { get; set; } = string.Empty;
        public string BugAnalysis { get; set; } = string.Empty;
        public TimeSpan TotalDuration { get; set; }
    }

    /// <summary>
    /// Voting Parallelizer - multiple evaluations for consensus
    /// </summary>
    public class VotingParallelizer
    {
        private readonly AnthropicClient _client;
        private readonly string _model;

        public VotingParallelizer(
            AnthropicClient client,
            string model = "claude-sonnet-4-20250514")
        {
            _client = client;
            _model = model;
        }

        /// <summary>
        /// Get multiple votes on a decision
        /// </summary>
        public async Task<VotingResult> VoteAsync(
            string question,
            List<string> options,
            int voterCount = 3)
        {
            var optionsList = string.Join("\n", options.Select((o, i) => $"{i + 1}. {o}"));
            var prompt = $@"Consider this question:
{question}

Options:
{optionsList}

Analyze carefully and respond with only the number of your chosen option.";

            var tasks = Enumerable.Range(0, voterCount).Select(async i =>
            {
                var messages = new List<Message>
                {
                    new Message(RoleType.User, prompt)
                };

                var parameters = new MessageParameters
                {
                    Messages = messages,
                    Model = _model,
                    MaxTokens = 10,
                    Temperature = 0.7 // Some variance for diverse opinions
                };

                var response = await _client.Messages.GetClaudeMessageAsync(parameters);
                var text = response.Content.First().Text.Trim();

                if (int.TryParse(text, out var vote) && vote >= 1 && vote <= options.Count)
                {
                    return vote - 1; // 0-indexed
                }
                return -1; // Invalid vote
            }).ToList();

            var votes = await Task.WhenAll(tasks);
            var validVotes = votes.Where(v => v >= 0).ToList();

            // Count votes
            var voteCounts = validVotes
                .GroupBy(v => v)
                .ToDictionary(g => g.Key, g => g.Count());

            var winningOption = voteCounts.Count > 0
                ? voteCounts.OrderByDescending(kv => kv.Value).First().Key
                : 0;

            return new VotingResult
            {
                WinningOption = options[winningOption],
                WinningIndex = winningOption,
                VoteCounts = options.Select((o, i) =>
                    new VoteCount { Option = o, Votes = voteCounts.GetValueOrDefault(i, 0) }).ToList(),
                TotalVotes = validVotes.Count,
                Consensus = voteCounts.Count > 0 &&
                    voteCounts.Values.Max() > (double)validVotes.Count / 2
            };
        }

        /// <summary>
        /// Safety voting - all voters must agree for approval
        /// </summary>
        public async Task<SafetyVotingResult> SafetyVoteAsync(
            string content,
            int voterCount = 3)
        {
            var prompt = $@"Evaluate if this content is safe and appropriate:

{content}

Respond with only 'SAFE' or 'UNSAFE'.";

            var tasks = Enumerable.Range(0, voterCount).Select(async i =>
            {
                var messages = new List<Message>
                {
                    new Message(RoleType.User, prompt)
                };

                var parameters = new MessageParameters
                {
                    Messages = messages,
                    Model = _model,
                    MaxTokens = 10
                };

                var response = await _client.Messages.GetClaudeMessageAsync(parameters);
                var text = response.Content.First().Text.Trim().ToUpper();

                return text.Contains("SAFE") && !text.Contains("UNSAFE");
            }).ToList();

            var votes = await Task.WhenAll(tasks);
            var safeVotes = votes.Count(v => v);

            return new SafetyVotingResult
            {
                IsSafe = votes.All(v => v), // Require unanimous agreement
                SafeVotes = safeVotes,
                UnsafeVotes = votes.Length - safeVotes,
                Unanimous = votes.All(v => v) || votes.All(v => !v)
            };
        }
    }

    public class VotingResult
    {
        public string WinningOption { get; set; } = string.Empty;
        public int WinningIndex { get; set; }
        public List<VoteCount> VoteCounts { get; set; } = new();
        public int TotalVotes { get; set; }
        public bool Consensus { get; set; }
    }

    public class VoteCount
    {
        public string Option { get; set; } = string.Empty;
        public int Votes { get; set; }
    }

    public class SafetyVotingResult
    {
        public bool IsSafe { get; set; }
        public int SafeVotes { get; set; }
        public int UnsafeVotes { get; set; }
        public bool Unanimous { get; set; }
    }

    /// <summary>
    /// Guardrails Parallelizer - run guardrails in parallel with main task
    /// </summary>
    public class GuardrailsParallelizer
    {
        private readonly AnthropicClient _client;
        private readonly string _model;

        public GuardrailsParallelizer(
            AnthropicClient client,
            string model = "claude-sonnet-4-20250514")
        {
            _client = client;
            _model = model;
        }

        /// <summary>
        /// Execute task with parallel guardrails
        /// </summary>
        public async Task<GuardrailedResult> ExecuteWithGuardrailsAsync(
            string input,
            string taskPrompt,
            List<string> guardrailPrompts)
        {
            // Run main task and guardrails in parallel
            var mainTask = ExecuteMainTaskAsync(taskPrompt);
            var guardrailTasks = guardrailPrompts.Select(async (prompt, index) =>
            {
                var checkPrompt = prompt.Replace("{input}", input);
                return await CheckGuardrailAsync(checkPrompt, $"guardrail_{index}");
            }).ToList();

            await Task.WhenAll(new[] { mainTask }.Concat(guardrailTasks.Cast<Task>()));

            var mainResult = await mainTask;
            var guardrailResults = await Task.WhenAll(guardrailTasks);

            var passed = guardrailResults.All(g => g.Passed);

            return new GuardrailedResult
            {
                Result = passed ? mainResult : null,
                Blocked = !passed,
                GuardrailResults = guardrailResults.ToList(),
                BlockingGuardrails = guardrailResults
                    .Where(g => !g.Passed)
                    .Select(g => g.Name)
                    .ToList()
            };
        }

        private async Task<string> ExecuteMainTaskAsync(string prompt)
        {
            var messages = new List<Message>
            {
                new Message(RoleType.User, prompt)
            };

            var parameters = new MessageParameters
            {
                Messages = messages,
                Model = _model,
                MaxTokens = 4096
            };

            var response = await _client.Messages.GetClaudeMessageAsync(parameters);
            return response.Content.First().Text;
        }

        private async Task<GuardrailResult> CheckGuardrailAsync(string prompt, string name)
        {
            var messages = new List<Message>
            {
                new Message(RoleType.User, prompt + "\n\nRespond with only 'PASS' or 'FAIL'.")
            };

            var parameters = new MessageParameters
            {
                Messages = messages,
                Model = "claude-3-haiku-20240307", // Faster model for guardrails
                MaxTokens = 10
            };

            var response = await _client.Messages.GetClaudeMessageAsync(parameters);
            var result = response.Content.First().Text.Trim().ToUpper();

            return new GuardrailResult
            {
                Name = name,
                Passed = result.Contains("PASS")
            };
        }
    }

    public class GuardrailResult
    {
        public string Name { get; set; } = string.Empty;
        public bool Passed { get; set; }
    }

    public class GuardrailedResult
    {
        public string? Result { get; set; }
        public bool Blocked { get; set; }
        public List<GuardrailResult> GuardrailResults { get; set; } = new();
        public List<string> BlockingGuardrails { get; set; } = new();
    }

    public static class ParallelizationExamples
    {
        public static async Task CodeReviewExample()
        {
            var apiKey = Environment.GetEnvironmentVariable("ANTHROPIC_API_KEY")
                ?? throw new InvalidOperationException("ANTHROPIC_API_KEY not set");

            var client = new AnthropicClient(apiKey);
            var parallelizer = new SectioningParallelizer(client);

            var code = @"
public class UserService
{
    private string connectionString;

    public User GetUser(int id)
    {
        var query = $""SELECT * FROM users WHERE id = {id}"";
        // Execute query...
        return null;
    }
}";

            var result = await parallelizer.ProcessCodeReviewAsync(code);

            Console.WriteLine("=== Code Review Results ===");
            Console.WriteLine($"\nSecurity:\n{result.SecurityAnalysis}");
            Console.WriteLine($"\nPerformance:\n{result.PerformanceAnalysis}");
            Console.WriteLine($"\nMaintainability:\n{result.MaintainabilityAnalysis}");
            Console.WriteLine($"\nPotential Bugs:\n{result.BugAnalysis}");
            Console.WriteLine($"\nTotal Duration: {result.TotalDuration.TotalMilliseconds}ms");
        }

        public static async Task VotingExample()
        {
            var apiKey = Environment.GetEnvironmentVariable("ANTHROPIC_API_KEY")
                ?? throw new InvalidOperationException("ANTHROPIC_API_KEY not set");

            var client = new AnthropicClient(apiKey);
            var voter = new VotingParallelizer(client);

            var result = await voter.VoteAsync(
                "What's the best approach for handling errors in an API?",
                new List<string>
                {
                    "Return HTTP status codes with error messages",
                    "Throw exceptions and let middleware handle them",
                    "Return a result object with success/failure status"
                },
                voterCount: 5);

            Console.WriteLine($"Winning option: {result.WinningOption}");
            Console.WriteLine($"Consensus reached: {result.Consensus}");
            foreach (var vc in result.VoteCounts)
            {
                Console.WriteLine($"  {vc.Option}: {vc.Votes} votes");
            }
        }
    }
}

File: skills/agent-patterns/templates/dart/parallelization.dart
/// Parallelization Pattern Implementation for Dart
/// Concurrent LLM calls with sectioning, voting, and guardrails

import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;

/// Anthropic API client
class AnthropicClient {
  final String apiKey;
  final http.Client httpClient;

  AnthropicClient({
    required this.apiKey,
    http.Client? httpClient,
  }) : httpClient = httpClient ?? http.Client();

  Future<String> createMessage({
    required String prompt,
    String model = 'claude-sonnet-4-20250514',
    int maxTokens = 4096,
  }) async {
    final response = await httpClient.post(
      Uri.parse('https://api.anthropic.com/v1/messages'),
      headers: {
        'x-api-key': apiKey,
        'anthropic-version': '2023-06-01',
        'content-type': 'application/json',
      },
      body: jsonEncode({
        'model': model,
        'max_tokens': maxTokens,
        'messages': [
          {'role': 'user', 'content': prompt}
        ],
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('API error (${response.statusCode}): ${response.body}');
    }

    final data = jsonDecode(response.body);
    final content = data['content'] as List;

    for (var block in content) {
      if (block['type'] == 'text') {
        return block['text'] as String;
      }
    }

    throw Exception('No text content in response');
  }
}

/// Result of processing a section
class SectionResult {
  final int index;
  final String section;
  final String result;
  final bool success;
  final String? error;

  SectionResult({
    required this.index,
    required this.section,
    required this.result,
    required this.success,
    this.error,
  });
}

/// Processes independent sections of work in parallel.
///
/// Example:
/// ```dart
/// final parallelizer = SectioningParallelizer(
///   client: client,
///   promptTemplate: (section) => 'Translate to Spanish: $section',
/// );
/// final results = await parallelizer.process([
///   'Hello, how are you?',
///   'The weather is nice today.',
///   'I love programming.',
/// ]);
/// ```
class SectioningParallelizer {
  final AnthropicClient client;
  final String model;
  final String Function(String section) promptTemplate;
  final int? maxConcurrency;

  SectioningParallelizer({
    required this.client,
    required this.promptTemplate,
    this.model = 'claude-sonnet-4-20250514',
    this.maxConcurrency,
  });

  /// Process all sections in parallel
  Future<List<SectionResult>> process(List<String> sections) async {
    if (maxConcurrency != null && maxConcurrency! > 0) {
      return _processWithLimit(sections, maxConcurrency!);
    }

    final futures = sections.asMap().entries.map((entry) async {
      final index = entry.key;
      final section = entry.value;
      try {
        final prompt = promptTemplate(section);
        final result = await client.createMessage(
          prompt: prompt,
          model: model,
        );
        return SectionResult(
          index: index,
          section: section,
          result: result,
          success: true,
        );
      } catch (e) {
        return SectionResult(
          index: index,
          section: section,
          result: '',
          success: false,
          error: e.toString(),
        );
      }
    }).toList();

    return Future.wait(futures);
  }

  Future<List<SectionResult>> _processWithLimit(
    List<String> sections,
    int limit,
  ) async {
    final results = <SectionResult>[];
    final chunks = _chunkList(sections.asMap().entries.toList(), limit);

    for (final chunk in chunks) {
      final chunkFutures = chunk.map((entry) async {
        final index = entry.key;
        final section = entry.value;
        try {
          final prompt = promptTemplate(section);
          final result = await client.createMessage(
            prompt: prompt,
            model: model,
          );
          return SectionResult(
            index: index,
            section: section,
            result: result,
            success: true,
          );
        } catch (e) {
          return SectionResult(
            index: index,
            section: section,
            result: '',
            success: false,
            error: e.toString(),
          );
        }
      }).toList();

      results.addAll(await Future.wait(chunkFutures));
    }

    // Sort by original index
    results.sort((a, b) => a.index.compareTo(b.index));
    return results;
  }

  List<List<T>> _chunkList<T>(List<T> list, int chunkSize) {
    final chunks = <List<T>>[];
    for (var i = 0; i < list.length; i += chunkSize) {
      final end = (i + chunkSize < list.length) ? i + chunkSize : list.length;
      chunks.add(list.sublist(i, end));
    }
    return chunks;
  }
}

/// Result of a vote
class VoteResult {
  final int index;
  final String response;
  final bool success;
  final String? error;

  VoteResult({
    required this.index,
    required this.response,
    required this.success,
    this.error,
  });
}

/// Voting result with aggregated responses
class VotingResult {
  final String winner;
  final int winnerCount;
  final Map<String, int> votes;
  final List<VoteResult> allResponses;

  VotingResult({
    required this.winner,
    required this.winnerCount,
    required this.votes,
    required this.allResponses,
  });

  double get confidence => allResponses.isEmpty
      ? 0.0
      : winnerCount / allResponses.where((r) => r.success).length;
}

/// Gets multiple LLM responses and aggregates via voting.
///
/// Example:
/// ```dart
/// final voter = VotingParallelizer(
///   client: client,
///   numVoters: 5,
///   extractAnswer: (response) => response.split('\n').first.toLowerCase(),
/// );
/// final result = await voter.vote('Is water wet? Answer yes or no.');
/// print('Winner: ${result.winner} (${result.confidence * 100}% confidence)');
/// ```
class VotingParallelizer {
  final AnthropicClient client;
  final String model;
  final int numVoters;
  final String Function(String response)? extractAnswer;

  VotingParallelizer({
    required this.client,
    this.model = 'claude-sonnet-4-20250514',
    this.numVoters = 3,
    this.extractAnswer,
  });

  /// Get votes and aggregate
  Future<VotingResult> vote(String prompt) async {
    final futures = List.generate(numVoters, (index) async {
      try {
        final response = await client.createMessage(
          prompt: prompt,
          model: model,
        );
        return VoteResult(
          index: index,
          response: response,
          success: true,
        );
      } catch (e) {
        return VoteResult(
          index: index,
          response: '',
          success: false,
          error: e.toString(),
        );
      }
    });

    final results = await Future.wait(futures);
    return _aggregateVotes(results);
  }

  VotingResult _aggregateVotes(List<VoteResult> results) {
    final votes = <String, int>{};

    for (final result in results) {
      if (!result.success) continue;

      final answer = extractAnswer != null
          ? extractAnswer!(result.response)
          : result.response.trim();

      votes[answer] = (votes[answer] ?? 0) + 1;
    }

    if (votes.isEmpty) {
      return VotingResult(
        winner: '',
        winnerCount: 0,
        votes: votes,
        allResponses: results,
      );
    }

    final sorted = votes.entries.toList()
      ..sort((a, b) => b.value.compareTo(a.value));

    return VotingResult(
      winner: sorted.first.key,
      winnerCount: sorted.first.value,
      votes: votes,
      allResponses: results,
    );
  }
}

/// Guardrail check result
class GuardrailResult {
  final String name;
  final bool passed;
  final String reason;
  final double? score;

  GuardrailResult({
    required this.name,
    required this.passed,
    required this.reason,
    this.score,
  });
}

/// Combined guardrails result
class GuardrailsResult {
  final bool allPassed;
  final List<GuardrailResult> results;
  final String? response;

  GuardrailsResult({
    required this.allPassed,
    required this.results,
    this.response,
  });

  List<GuardrailResult> get failed =>
      results.where((r) => !r.passed).toList();
}

/// A guardrail definition
class Guardrail {
  final String name;
  final String prompt;
  final bool Function(String response) check;

  Guardrail({
    required this.name,
    required this.prompt,
    required this.check,
  });
}

/// Runs guardrail checks in parallel with the main task.
///
/// Example:
/// ```dart
/// final parallelizer = GuardrailsParallelizer(
///   client: client,
///   taskPrompt: 'Write code to sort a list',
///   guardrails: [
///     Guardrail(
///       name: 'safe_code',
///       prompt: 'Does this request safe code generation?',
///       check: (r) => r.toLowerCase().contains('yes'),
///     ),
///   ],
/// );
/// final result = await parallelizer.execute('Sort this list: [3, 1, 2]');
/// ```
class GuardrailsParallelizer {
  final AnthropicClient client;
  final String model;
  final String taskPrompt;
  final List<Guardrail> guardrails;
  final bool stopOnFailure;

  GuardrailsParallelizer({
    required this.client,
    required this.taskPrompt,
    required this.guardrails,
    this.model = 'claude-sonnet-4-20250514',
    this.stopOnFailure = true,
  });

  /// Execute task with parallel guardrails
  Future<GuardrailsResult> execute(String input) async {
    // Run guardrails and main task in parallel
    final guardrailFutures = guardrails.map((g) => _checkGuardrail(g, input));
    final taskFuture = client.createMessage(
      prompt: '$taskPrompt\n\nInput: $input',
      model: model,
    );

    // Wait for all to complete
    final guardrailResults = await Future.wait(guardrailFutures);
    final taskResponse = await taskFuture;

    final allPassed = guardrailResults.every((r) => r.passed);

    return GuardrailsResult(
      allPassed: allPassed,
      results: guardrailResults,
      response: (!stopOnFailure || allPassed) ? taskResponse : null,
    );
  }

  Future<GuardrailResult> _checkGuardrail(
    Guardrail guardrail,
    String input,
  ) async {
    try {
      final response = await client.createMessage(
        prompt: '${guardrail.prompt}\n\nContent: $input\n\nRespond with yes or no and a brief reason.',
        model: model,
        maxTokens: 256,
      );

      final passed = guardrail.check(response);
      return GuardrailResult(
        name: guardrail.name,
        passed: passed,
        reason: response,
      );
    } catch (e) {
      return GuardrailResult(
        name: guardrail.name,
        passed: false,
        reason: 'Error: $e',
      );
    }
  }
}

/// Code review result
class CodeReviewResult {
  final List<ReviewCategory> categories;
  final bool allPassed;
  final double overallScore;

  CodeReviewResult({
    required this.categories,
    required this.allPassed,
    required this.overallScore,
  });
}

/// A category in a code review
class ReviewCategory {
  final String name;
  final double score;
  final List<String> issues;

  ReviewCategory({
    required this.name,
    required this.score,
    required this.issues,
  });

  bool get passed => score >= 0.7;
}

/// Parallel code reviewer
class ParallelCodeReviewer {
  final AnthropicClient client;
  final String model;

  ParallelCodeReviewer({
    required this.client,
    this.model = 'claude-sonnet-4-20250514',
  });

  /// Review code across multiple aspects in parallel
  Future<CodeReviewResult> review(String code) async {
    final aspects = [
      ('security', 'Review for security vulnerabilities'),
      ('performance', 'Review for performance issues'),
      ('maintainability', 'Review for code quality and maintainability'),
      ('correctness', 'Review for logical errors and bugs'),
    ];

    final futures = aspects.map((aspect) async {
      final name = aspect.$1;
      final description = aspect.$2;

      final prompt = '''
$description in this code:

```
$code
```

Respond in JSON format:
{
  "score": 0.0-1.0,
  "issues": ["issue1", "issue2"]
}
''';

      try {
        final response = await client.createMessage(
          prompt: prompt,
          model: model,
          maxTokens: 512,
        );

        final jsonStr = _extractJson(response);
        final data = jsonDecode(jsonStr);

        return ReviewCategory(
          name: name,
          score: (data['score'] as num).toDouble(),
          issues: (data['issues'] as List).cast<String>(),
        );
      } catch (e) {
        return ReviewCategory(
          name: name,
          score: 0.0,
          issues: ['Error during review: $e'],
        );
      }
    });

    final categories = await Future.wait(futures);
    final allPassed = categories.every((c) => c.passed);
    final overallScore = categories.isEmpty
        ? 0.0
        : categories.map((c) => c.score).reduce((a, b) => a + b) /
            categories.length;

    return CodeReviewResult(
      categories: categories,
      allPassed: allPassed,
      overallScore: overallScore,
    );
  }

  String _extractJson(String text) {
    final start = text.indexOf('{');
    final end = text.lastIndexOf('}');
    if (start >= 0 && end > start) {
      return text.substring(start, end + 1);
    }
    return text;
  }
}

// Example usage
Future<void> exampleParallelization() async {
  final apiKey = const String.fromEnvironment('ANTHROPIC_API_KEY');
  if (apiKey.isEmpty) {
    throw Exception('ANTHROPIC_API_KEY environment variable not set');
  }

  final client = AnthropicClient(apiKey: apiKey);

  // Sectioning parallelization
  print('=== Sectioning Parallelization ===');
  final parallelizer = SectioningParallelizer(
    client: client,
    promptTemplate: (section) => 'Translate to French: $section',
    maxConcurrency: 3,
  );

  final sections = [
    'Hello, how are you?',
    'The weather is nice today.',
    'I love programming.',
    'Dart is a great language.',
  ];

  final results = await parallelizer.process(sections);
  for (final result in results) {
    print('${result.section} -> ${result.result}');
  }

  // Voting parallelization
  print('\n=== Voting Parallelization ===');
  final voter = VotingParallelizer(
    client: client,
    numVoters: 5,
    extractAnswer: (response) {
      final lower = response.toLowerCase();
      if (lower.contains('yes')) return 'yes';
      if (lower.contains('no')) return 'no';
      return response.trim();
    },
  );

  final voteResult = await voter.vote('Is the sky blue? Answer yes or no.');
  print('Winner: ${voteResult.winner}');
  print('Confidence: ${(voteResult.confidence * 100).toStringAsFixed(0)}%');
  print('Votes: ${voteResult.votes}');

  // Guardrails parallelization
  print('\n=== Guardrails Parallelization ===');
  final guardrailed = GuardrailsParallelizer(
    client: client,
    taskPrompt: 'Write a function based on this request:',
    guardrails: [
      Guardrail(
        name: 'safe_request',
        prompt: 'Is this a safe, non-malicious code request?',
        check: (r) => r.toLowerCase().contains('yes'),
      ),
      Guardrail(
        name: 'appropriate',
        prompt: 'Is this request appropriate for a coding assistant?',
        check: (r) => r.toLowerCase().contains('yes'),
      ),
    ],
  );

  final guardrailResult = await guardrailed.execute('Sort a list of numbers');
  print('All guardrails passed: ${guardrailResult.allPassed}');
  if (guardrailResult.response != null) {
    print('Response: ${guardrailResult.response!.substring(0, 100)}...');
  }

  // Code review
  print('\n=== Parallel Code Review ===');
  final reviewer = ParallelCodeReviewer(client: client);
  final reviewResult = await reviewer.review('''
def process_user_input(user_input):
    query = f"SELECT * FROM users WHERE name = '{user_input}'"
    return execute_query(query)
''');

  print('Overall score: ${(reviewResult.overallScore * 100).toStringAsFixed(0)}%');
  print('All passed: ${reviewResult.allPassed}');
  for (final category in reviewResult.categories) {
    print('${category.name}: ${(category.score * 100).toStringAsFixed(0)}%');
    for (final issue in category.issues) {
      print('  - $issue');
    }
  }
}

void main() async {
  try {
    await exampleParallelization();
  } catch (e) {
    print('Error: $e');
  }
}
